src/launchpad/artifacts/android/aab.py

```diff
@@ -75,9 +77,25 @@ def get_primary_apks(self, device_spec: DeviceSpec = DeviceSpec()) -> list[APK]:
 
             apks = []
             for apk_path in apks_dir.glob("*.apk"):
-                apks.append(APK(apk_path))
+                apks.append(APK(apk_path, self.get_dex_mapping()))
 
             self._primary_apks = apks
             return apks
         finally:
             cleanup_directory(apks_dir)
+
+    def get_dex_mapping(self) -> DexMapping | None:
+        if self._dex_mapping is not None:
+            return self._dex_mapping
+
+        dex_mapping_files = list(self._extract_dir.rglob("proguard.map"))
+        if len(dex_mapping_files) > 1:
+            raise ValueError("Multiple dex_mapping.txt files found in AAB")
```

Comment:
One thing I've been meaning to do at some point is create error subclasses instead of using ValueError everywhere, I believe that's the idiomatic way similar to Ruby.

Reply:
We can add a follow up to do better error auditing. I think we should think through best practice here and move forward once we decide on the best plan.